lib/granite/context_proxy.rb

```diff
@@ -0,0 +1,39 @@
+require 'granite/context_proxy/data'
+require 'granite/context_proxy/proxy'
+
+module Granite
+  # This concern contains class methods used for actions and projectors
+  #
+  module ContextProxy
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def using(data)
+        Proxy.new(self, Data.wrap(data))
+      end
+
+      def as(performer)
+        using(performer: performer)
+      end
+
+      def with_context(context)
+        key = proxy_context_key
+        old_context = Thread.current[key]
+        Thread.current[key] = context
+        yield
+      ensure
+        Thread.current[key] = old_context
+      end
+
+      def proxy_context
+        Thread.current[proxy_context_key]
+      end
+
+      private
+
+      def proxy_context_key
+        :"granite_proxy_performer_#{hash}"
```

Comment:
I don't exactly understand why this `proxy_context_key` needs to be dynamic. Do we have a test for that use-case?

Reply:
It's less of a problem with `BA.hash` but just wondering.

Reply:
when you call `provide_book.performer` you are only executing one method at the same time which is proxied through the Proxy wrapper. I think if you implement a static value of the key and execute this example, it will still pass. On the other hand what could fail is:

```
get_book = GetBook.new(library) # notice the lack of as(...)
get_book.provide_book.method_which_takes_a_block do
  get_book.performer # I think here with a static implementation we would get wrong result (provide_book.performer) instead of nil
end
```

but taking this to the extreme level where:
* a class invokes the same class
* the parent object does not have a context set thus there is no proxying behavior

then I guess even the current solution will return the wrong result, am I correct?

Reply:
Actually I think I was wrong. BA always sets it's `ctx` during initialization in https://github.com/toptal/granite/pull/76/files#diff-6edf00ef9c1be828de83ded2a18bae8a9b63423df3937ebca54bb60971ad924eR16.

So even if we initialize the action in initializer, e.g.:
```ruby
class GetBook
  def initialize
    super
    @provide_book = ProvideBook.as(librarian).new
  end
end

GetBook.as(customer).new
```

It should:
1. Set `Thread.current[key]` to `{performer: customer}`
2. Initialize `GetBook`
3. Set `get_book.ctx` to `{performer: customer}`
4. Set `Thread.current[key]` to `{performer: librarian}`
5. Initialize `ProvideBook`
6. Set `provide_book.ctx` to `{performer: librarian}`
7. Revert `Thread.current[key]` to `{performer: customer}`
8. Revert `Thread.current[key]` to `nil`

Even if we initialize `provide_book` before calling `super` it just moves step 3 between 7 & 8, so it's still set correctly. So I think we can get away with using same key for every action.

Reply:
I've updated the PR to use same key for all contexts.